lib/ts/AcidPtr.h

```diff
@@ -0,0 +1,198 @@
+      m.unlock();
+    }
+  }
+
+private:
+  std::vector<Mutex_t> mutexes;
+
+  /// use the other constructor to define how many locks you want.
+  LockPool() = delete;
+};
+
+template <typename T> class AcidCommitPtr;
+template <typename T> class AcidPtr;
+
+using AcidPtrMutex = std::mutex; // TODO: use shared_mutex when available
+using AcidPtrLock  = std::unique_lock<AcidPtrMutex>;
+AcidPtrMutex &AcidPtrMutexGet(void const *ptr); // used for read, and write swap
+
+using AcidCommitMutex = std::mutex;
+using AcidCommitLock  = std::unique_lock<AcidCommitMutex>;
+AcidCommitMutex &AcidCommitMutexGet(void const *ptr); // used for write block
+
+///////////////////////////////////////////
+/// AcidPtr
+/** just a thread safe shared pointer.
+ *
+ */
+
+template <typename T> class AcidPtr
+{
+private:
+  std::shared_ptr<T> data_ptr;
+
+public:
+  AcidPtr(const AcidPtr &) = delete;
+  AcidPtr &operator=(const AcidPtr &) = delete;
+
+  AcidPtr() : data_ptr(new T()) {}
+  AcidPtr(T *data) : data_ptr(data) {}
+
+  const std::shared_ptr<const T>
+  getPtr() const
+  { // wait until we have exclusive pointer access.
+    auto ptr_lock = AcidPtrLock(AcidPtrMutexGet(&data_ptr));
+    // copy the pointer
+    return data_ptr;
+    //[end scope] unlock ptr_lock
+  }
+
+  void
+  commit(T *data)
+  {
+    // wait until existing commits finish, avoid race conditions
+    auto commit_lock = AcidCommitLock(AcidCommitMutexGet(&data_ptr));
+    // wait until we have exclusive pointer access.
+    auto ptr_lock = AcidPtrLock(AcidPtrMutexGet(&data_ptr));
+    // overwrite the pointer
+    data_ptr.reset(data);
+    //[end scope] unlock commit_lock & ptr_lock
+  }
+
+  AcidCommitPtr<T>
+  startCommit()
+  {
+    return AcidCommitPtr<T>(*this);
+  }
+
+  friend class AcidCommitPtr<T>;
+
+protected:
+  void
+  _finishCommit(T *data)
+  {
+    // wait until we have exclusive pointer access.
+    auto ptr_lock = AcidPtrLock(AcidPtrMutexGet(&data_ptr));
+    // overwrite the pointer
+    data_ptr.reset(data);
+    //[end scope] unlock ptr_lock
+  }
+};
+
+///////////////////////////////////////////
+/// AcidCommitPtr
+
+/// a globally exclusive pointer, for commiting changes to AcidPtr.
+/** used for COPY_SWAP functionality.
+ * 1. copy data (construct)
+ * 2. overwrite data (scope)
+ * 3. update live data pointer (destruct)
+ */
+template <typename T> class AcidCommitPtr : public std::unique_ptr<T>
+{
+private:
+  AcidCommitLock commit_lock; // block other writers from starting
+  AcidPtr<T> &data;           // data location
+
+public:
+  AcidCommitPtr()                      = delete;
+  AcidCommitPtr(const AcidCommitPtr &) = delete;
+  AcidCommitPtr &operator=(const AcidCommitPtr<T> &) = delete;
+
+  AcidCommitPtr(AcidPtr<T> &data_ptr) : commit_lock(AcidCommitMutexGet(&data_ptr)), data(data_ptr)
+  {
+    // wait for exclusive commit access to the data
+    // copy the data to new memory
+    std::unique_ptr<T>::reset(new T(*data.getPtr()));
```

Comment:
What does this do? Allocate and then immediately destroy? Is there side effect that's important? Why use `std::unique_ptr` instead of
```
T(*data.getPtr());
```
?

Reply:
it is effectively `this->reset(new T(*data.getPtr()));` 
1. copy the data with `new T(*data.getPtr())`
2. update the commit pointer with `reset()`